chat_example/src/main/java/io/skygear/chatexample/ApiParamsActivity.kt

```diff
@@ -0,0 +1,130 @@
+
+class ApiParamsActivity : AppCompatActivity() {
+
+    private var paramsAdapter: ApiParamsAdapter? = null
+    private var usersAdapter: ChatUsesAdapter? = null
+    val mSkygear: Container = Container.defaultContainer(this)
+    val mChatContainer: ChatContainer
+
+    init{
+        mChatContainer = ChatContainer.getInstance(mSkygear)
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_api_params)
+        title = fetchExtras().name
+
+        initViews()
+        initParams(fetchExtras())
+    }
+
+    private fun initViews() {
+        continue_btn.setOnClickListener { submitParams() }
+    }
+
+    private fun fetchExtras(): ApiTask {
+        return intent.getSerializableExtra(EXTRAS_KEY) as ApiTask
+    }
+
+    private fun initParams(apiTask: ApiTask) {
+        if(apiTask.loginRequired) {
+            // Pick User IDs
+            if (!Utils.isLoggedIn(this)) {
+                Toast.makeText(this, "Please login first!", Toast.LENGTH_SHORT).show()
+                finish()
+                return
+            }
+        }
+
+        if(apiTask.params.isEmpty()) {
+            val intent = Intent(this, ApiTestActivity::class.java)
+            intent.putExtra(EXTRAS_KEY, apiTask)
+            startActivity(intent)
+            finish()
+            return
+        }
+
+        initRecyclerView(apiTask)
+    }
+
+    private fun initRecyclerView(apiTask: ApiTask) {
+        recycler_view.layoutManager = LinearLayoutManager(this)
+
+        if(apiTask.params.containsKey(USER_ID_KEY)) {
+            usersAdapter = ChatUsesAdapter(mSkygear.auth.currentUser.id)
+            ParticipantsFetcher(mSkygear).fetch(object : GetParticipantsCallback {
+                override fun onGetCachedResult(participantsMap: MutableMap<String, Participant>?) {
+                    usersAdapter?.setParticipants(participantsMap)
+                }
+
+                override fun onSuccess(participantsMap: MutableMap<String, Participant>?) {
+                    usersAdapter?.setParticipants(participantsMap)
+                }
+
+
+                override fun onFail(error: Error) {
+                }
+            })
+        } else {
+            // Input params
+            paramsAdapter = ApiParamsAdapter(this, apiTask)
+        }
+        recycler_view.adapter = paramsAdapter ?: usersAdapter
+    }
+
+    private fun submitParams() {
+        if(paramsAdapter != null) {
+            val finalAdapter = paramsAdapter
+            if (finalAdapter != null) {
+                if (!finalAdapter.isAllParamsFilled()) {
+                    Toast.makeText(this, R.string.params_incomplete, Toast.LENGTH_SHORT).show()
+                    return
+                }
+                val intent = Intent(this, ApiTestActivity::class.java)
+                intent.putExtra(EXTRAS_KEY, finalAdapter.getParams())
+                startActivity(intent)
+                finish()
```

Comment:
It would be better if we can stay in the original screen if the request is fail, for example login fail. We may want to correct the data and re-login again.